component/stream/read_write_filename.go

```diff
@@ -0,0 +1,478 @@
+	handle, err := rw.NextComponent().OpenFile(options)
+	if err != nil {
+		log.Err("Stream::OpenFile : error failed to open file %s [%s]", options.Name, err.Error())
+		return handle, err
+	}
+	if !rw.StreamOnly {
+		err = rw.createFileCache(handle)
+		if err != nil {
+			log.Err("Stream::OpenFile : error failed to create cache object %s [%s]", options.Name, err.Error())
+		}
+	}
+	return handle, err
+}
+
+func (rw *ReadWriteFilenameCache) ReadInBuffer(options internal.ReadInBufferOptions) (int, error) {
+	// log.Trace("Stream::ReadInBuffer : name=%s, handle=%d, offset=%d", options.Handle.Path, options.Handle.ID, options.Offset)
+	if !rw.StreamOnly && options.Handle.CacheObj.StreamOnly {
+		err := rw.createFileCache(options.Handle)
+		if err != nil {
+			log.Err("Stream::ReadInBuffer : error failed to create cache object  %s [%s]", options.Handle.Path, err.Error())
+			return 0, err
+		}
+	}
+	if rw.StreamOnly || options.Handle.CacheObj.StreamOnly {
+		data, err := rw.NextComponent().ReadInBuffer(options)
+		if err != nil && err != io.EOF {
+			log.Err("Stream::ReadInBuffer : error failed to download requested data for %s: [%s]", options.Handle.Path, err.Error())
+		}
+		return data, err
+	}
+	if atomic.LoadInt64(&options.Handle.CacheObj.Size) == 0 {
+		return 0, nil
+	}
+	read, err := rw.readWriteBlocks(options.Handle, options.Offset, options.Data, false)
+	if err != nil {
+		log.Err("Stream::ReadInBuffer : error failed to download requested data for %s: [%s]", options.Handle.Path, err.Error())
+	}
+	return read, err
+}
+
+func (rw *ReadWriteFilenameCache) WriteFile(options internal.WriteFileOptions) (int, error) {
+	// log.Trace("Stream::WriteFile : name=%s, handle=%d, offset=%d", options.Handle.Path, options.Handle.ID, options.Offset)
+	if !rw.StreamOnly && options.Handle.CacheObj.StreamOnly {
+		err := rw.createFileCache(options.Handle)
+		if err != nil {
+			log.Err("Stream::WriteFile : error failed to create cache object %s [%s]", options.Handle.Path, err.Error())
+			return 0, err
+		}
+	}
+	if rw.StreamOnly || options.Handle.CacheObj.StreamOnly {
+		data, err := rw.NextComponent().WriteFile(options)
+		if err != nil && err != io.EOF {
+			log.Err("Stream::WriteFile : error failed to write data to %s: [%s]", options.Handle.Path, err.Error())
+		}
+		return data, err
+	}
+	written, err := rw.readWriteBlocks(options.Handle, options.Offset, options.Data, true)
+	if err != nil {
+		log.Err("Stream::WriteFile : error failed to write data to %s: [%s]", options.Handle.Path, err.Error())
+	}
+	options.Handle.Flags.Set(handlemap.HandleFlagDirty)
+	return written, err
+}
+
+// TODO: truncate in cache
+func (rw *ReadWriteFilenameCache) TruncateFile(options internal.TruncateFileOptions) error {
+	// log.Trace("Stream::TruncateFile : name=%s, size=%d", options.Name, options.Size)
+	if !rw.StreamOnly {
+		rw.purge(options.Name, false)
+	}
+	err := rw.NextComponent().TruncateFile(options)
```

Comment:
Shouldn't the order be reverse, like you update on container first and if it's successful we shall make changes in local cache.